tools/DeviceExplorer/DeviceExplorer/DevicesProcessor.cs

```diff
@@ -6,78 +6,157 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Text;
 using System.Threading.Tasks;
+using System.Windows.Forms;
+using Newtonsoft.Json.Linq;
 
 namespace DeviceExplorer
 {
     class DevicesProcessor
     {
-        private List<DeviceEntity> listOfDevices;
         private RegistryManager registryManager;
         private String iotHubConnectionString;
         private int maxCountOfDevices;
         private String protocolGatewayHostName;
 
         public DevicesProcessor(string iotHubConnenctionString, int devicesCount, string protocolGatewayName)
         {
-            this.listOfDevices = new List<DeviceEntity>();
             this.iotHubConnectionString = iotHubConnenctionString;
             this.maxCountOfDevices = devicesCount;
             this.protocolGatewayHostName = protocolGatewayName;
             this.registryManager = RegistryManager.CreateFromConnectionString(iotHubConnectionString);
         }
 
-        public async Task<List<DeviceEntity>> GetDevices()
+        /// <summary>
+        /// Receive an approximation of the devices in the IoT hub, up to the 
+        /// specified max.  This is not a definitive list, and seems to return
+        /// only up to around 1000 devices regardless of what you specify as 
+        /// the max.
+        /// </summary>
+        /// <remarks>
+        /// The GetDevicesAsync() method is currently marked as obsolete, but 
+        /// loading one thousand devices via this method is much, much faster
+        /// than doing so through the GetDeviceSample() method.  The reason for 
+        /// this is that that method makes one request to get the first page of
+        /// twins, and then another request for EVERY one of the devices in order
+        /// to get their keys and such (totaling 1001 requests for 1000 devices).  
+        /// If the call to FetchDeviceFromTwin() was removed from GetDevicesAsync(), 
+        /// it would be just as fast as GetDevicesAsync(), but it would not provide 
+        /// the device keys and other information to display in the table.
+        /// </remarks>
+        public async Task<List<DeviceEntity>> GetABunchOfDevices()
         {
-            try
+            var listOfDevices = new List<DeviceEntity>();
+            IEnumerable<Device> devices = await registryManager.GetDevicesAsync(maxCountOfDevices);
```

Comment:
This is an obsolete API that we will remove soon from the service side (so it won't work...)
The service team recommended us to avoid getting the identity (keys) if not required. You could change the list to exclude that information (and so avoid making the 1001 requests). The keys can be available only when right-click (or the details page is open).